### PR TITLE
Initial MSI credential implementation

### PR DESCRIPTION
**Pull request by schaabs**

### CODE REVIEW

File: sdk/identity/Azure.Identity/src/AzureCredential.cs [hunk 1]
```diff
@@ -8,6 +8,16 @@
 
 namespace Azure.Identity
 {
+    public class TokenResponse
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/AzureCredential.cs]: Separate file

Comment (pakrym) [reply sdk/identity/Azure.Identity/src/AzureCredential.cs]: Who uses this?

**Comment by pakrym** [comment]
As I understand live tests are hard to write for MSI, can we add mocking tests for this implementation?

**Review by AlexGhiondea — APPROVED** [review]

File: sdk/identity/Azure.Identity/src/IdentityClient.cs [hunk 2]
```diff
@@ -20,6 +20,8 @@ internal class IdentityClient
     {
         private readonly IdentityClientOptions _options;
         private readonly HttpPipeline _pipeline;
+        private readonly Uri ImdsEndptoint = new Uri("http://169.254.169.254/metadata/identity/oauth2/token");
```
Comment (jsquire) [inline sdk/identity/Azure.Identity/src/IdentityClient.cs]: Should we consider allowing this to be overridden by environment or options?

**Review by jsquire — APPROVED** [review]

File: sdk/identity/Azure.Identity/tests/Mock/MockToken.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Azure.Identity.Tests.Mock
+{
+    class MockToken
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/tests/Mock/MockToken.cs]: nit: access modifier

File: sdk/identity/Azure.Identity/tests/TokenCredentialProviderTests.cs [hunk 4]
```diff
@@ -71,23 +71,23 @@ public async Task CredentialSequenceValid()
...
             var provider = new AggregateCredential(cred1, cred2, cred3, cred4);
 
-            Assert.Equal("tokenA", await provider.GetTokenAsync(new string[] { "scopeA" }));
-            Assert.Equal("tokenB", await provider.GetTokenAsync(new string[] { "scopeB" }));
-            Assert.Equal("tokenC", await provider.GetTokenAsync(new string[] { "scopeC" }));
-            await Assert.ThrowsAsync<InvalidOperationException>(async () => await provider.GetTokenAsync(new string[] { "scopeD" }));
+            Assert.AreEqual("tokenA", await provider.GetTokenAsync(new string[] { "scopeA" }));
+            Assert.AreEqual("tokenB", await provider.GetTokenAsync(new string[] { "scopeB" }));
+            Assert.AreEqual("tokenC", await provider.GetTokenAsync(new string[] { "scopeC" }));
+            // await Assert.ThrowsAsync<InvalidOperationException>(async () => await provider.GetTokenAsync(new string[] { "scopeD" }));
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/tests/TokenCredentialProviderTests.cs]: Is this commented out by accident?

**Comment by AlexGhiondea** [comment]
@schaabs can this be merged?

**Pull request merged**